### PR TITLE
[bug] NF-32 홈 요약 조회와 DB trigger 테스트 안정화

### LINKED ISSUE #44: [bug] NF-32 홈 요약 조회와 DB trigger 테스트 안정화
## 관련 키 / 참고 링크
- 트래킹 키: `NF-32`
- Jira: https://parkjaehong.atlassian.net/browse/NF-32
- 관련 사후 리뷰:
  - #16 홈 테스트 현재 월 의존: https://github.com/SaGongSa-404/404_BE/pull/16#discussion_r3193630099
  - #16 월별 집계 인덱스 비효율: https://github.com/SaGongSa-404/404_BE/pull/16#discussion_r3193636534
  - #18 self_check_response_sets delete 정책: https://github.com/SaGongSa-404/404_BE/pull/18#discussion_r3194994706
  - #18 post_votes update guard: https://github.com/SaGongSa-404/404_BE/pull/18#discussion_r3195001074
  - #18 post_votes DELETE 테스트: https://github.com/SaGongSa-404/404_BE/pull/18#discussion_r3195005641

## 증상
- 홈 요약 테스트가 고정된 `BASE_TIME`과 `YearMonth.now(Asia/Seoul)`을 섞어 사용한다.
- 월별 rational choice 집계가 `to_char(decided_at at time zone ?, 'YYYY-MM') = ?` 형태라 인덱스 활용에 불리하다.
- self_check_response_sets 삭제 시 decision summary가 남을 수 있다.
- post_votes UPDATE trigger가 집계 관련 필드가 바뀌지 않아도 feed_posts.updated_at을 갱신할 수 있다.
- post_votes DELETE trigger branch가 테스트로 고정되어 있지 않다.

### 예상 결과
- 홈 요약 테스트는 실행 날짜와 무관하게 안정적으로 통과한다.
- 월별 집계는 사용자 timezone 기준 월 시작/끝 Instant 범위 조건으로 조회한다.
- self-check 하위 데이터 삭제 정책을 안전하게 반영한다.
- post vote counter trigger는 실제 집계 변경이 있을 때만 count/updated_at을 조정한다.
- DELETE branch도 smoke test로 검증한다.

### 실제 결과
- 2026년 5월 이후 실행 시 2026년 4월 decision 데이터와 현재 월 budget이 어긋나 rationalChoiceRate 기대값이 깨질 수 있다.
- 데이터가 쌓이면 홈 요약 조회가 해당 사용자의 decision을 더 넓게 스캔할 수 있다.
- trigger 동작이 실제 의미 없는 변경에도 feed_posts updated_at을 변경할 수 있다.

## 영향 및 발생 빈도 (Impact)
- 홈 요약은 자주 호출되는 API라 쿼리 안정성이 중요하다.
- trigger drift는 데이터 정합성과 테스트 신뢰도에 영향을 준다.

## 재현 경로
1. 현재 월과 다른 고정 decision 날짜로 HomeSummaryIntegrationTest를 실행한다.
2. 월별 rational choice query plan을 확인한다.
3. post_votes의 단순 metadata update와 delete를 수행해 feed_posts count/updated_at 변화를 확인한다.

## 환경
- [x] Local (내 컴퓨터)
- [x] Dev/Staging (테스트 서버)
- [ ] Prod (운영 서버)

## 심각도(Severity)
- [ ] S1 (서비스 중단/데이터 손상)
- [ ] S2 (핵심 기능 장애)
- [x] S3 (부분 장애/우회 가능)
- [ ] S4 (경미)

## 작업 범위
- 홈 요약 테스트의 날짜 의존 제거
- 월별 rational choice 집계 range query 전환
- self_check_response_sets delete 정책 반영 또는 차단
- post_votes trigger update guard 추가
- post_votes DELETE branch smoke test 추가

## Out of Scope
- mascot reaction 만료 정책 결정은 `NF-33`에서 별도 확인
- 홈 화면 신규 응답 필드 추가

### PR DESCRIPTION
**Pull request by PHJ2000**
# 🐛 Bugfix PR

## 🔗 작업 키 / 이슈 링크 (필수)
- Tracking Key: **NF-32**
- Jira: https://parkjaehong.atlassian.net/browse/NF-32

> PR 제목: `[bug] NF-32 홈 요약 조회와 DB trigger 테스트 안정화`  
> 브랜치: `fix/NF-32-home-trigger-test-hardening`

---

### 🔒 Close Issue (필수)
- GitHub: Closes #44

---

### 어떤 버그였나?
- 홈 요약 테스트가 고정 날짜와 현재 월을 섞어 사용해 실행 월에 따라 실패할 수 있었습니다.
- rational choice 월별 집계가 `to_char` 비교라 인덱스 활용에 불리했습니다.
- self_check_response_sets 삭제와 post_votes trigger edge case가 테스트로 고정되지 않았습니다.

### 왜 발생했나? (Root Cause)
- 테스트 fixture 기준 시간이 현재 월 계산과 분리되어 있었습니다.
- 월별 조회 조건이 timezone 변환 문자열 비교로 작성되어 range scan에 맞지 않았습니다.
- DB trigger 정책이 DELETE/metadata-only UPDATE branch까지 smoke test로 검증되지 않았습니다.

### 어떻게 고쳤나?
- HomeSummaryIntegrationTest fixture를 현재 월 기준으로 정리했습니다.
- rational choice 집계를 사용자 timezone 기준 month start/end range query로 변경했습니다.
- self_check_response_sets 삭제 차단 trigger와 post_votes metadata-only update guard를 신규 V3 migration으로 추가하고 smoke test를 보강했습니다.

---

### 재현 방법(가능하면)
- Steps: 현재 월과 다른 고정 decision 날짜로 HomeSummaryIntegrationTest를 실행합니다.
- 기대 결과: 실행 날짜와 무관하게 현재 월 fixture 기준으로 안정적으로 통과합니다.
- 실제 결과: 기존 테스트는 2026년 4월 fixture와 현재 월 budget이 어긋날 수 있었습니다.

### 재발 방지(있다면)
- [x] 회귀 테스트 추가
- [ ] 모니터링/알람 보강
- [ ] 런북/문서 보강

---

## ✅ Acceptance Criteria (AC) (필수)
- [x] AC1: 홈 요약 테스트의 실행 날짜 의존성을 제거한다.
- [x] AC2: rational choice 월 집계를 range query로 전환한다.
- [x] AC3: self-check 삭제 차단과 post_votes trigger edge case를 smoke test로 고정한다.

---

### 로컬
- Command: `./gradlew.bat testClasses`
- Result: PASS
- Command: `./gradlew.bat test --tests "com.sagongsa.backend.home.HomeSummaryIntegrationTest" --tests "com.sagongsa.backend.domain.DomainSchemaSmokeTest"`
- Result: PASS

### CI
- 링크/결과: 자동 첨부 확인 예정

### Smoke / 수동 검증
- 시나리오: 홈 요약 현재 월 집계, self_check_response_sets delete 차단, post_votes update/delete trigger 검증
- 결과: 통합 테스트로 검증

### 테스트 공백(있다면 이유 필수)
- mascot reaction 만료 정책 자체는 NF-33 정책 확인 범위로 남겼습니다.

---

## 🧭 영향 범위 (필수)
- [ ] API 스펙 변경 없음
- [x] API 스펙 변경 있음 (요약: 응답 필드 변경은 없고 홈 요약 집계 조건을 range query로 변경)
- [ ] DB 스키마 변경 없음
- [x] DB 스키마 변경 있음 (마이그레이션/락/시간: `V3__home_trigger_test_hardening.sql`, trigger/function 추가 및 교체)
- [x] 설정/환경변수 변경 없음
- [ ] 설정/환경변수 변경 있음 (항목: )
- [x] 캐시/큐/외부 연동 영향 없음
- [ ] 캐시/큐/외부 연동 영향 있음 (요약: )

---

## 💥 Breaking / Compatibility (필수)
- [x] Breaking change 없음
- [ ] Breaking change 있음 → 무엇이 깨지는지 / 마이그레이션 / 롤아웃 전략:

---

### Feature Flag
- [x] 사용 안함
- [ ] 사용함 → Flag/기본값/롤아웃:

### 모니터링/알림
- 확인할 대시보드/지표: home summary latency, migration 성공 여부, trigger 오류 로그
- 에러/로그 키워드: `prevent_self_check_response_set_delete`, `sync_feed_post_vote_counts`

---

### 리스크
- trigger/function 변경이 포함되어 있으므로 migration 적용 순서와 기존 DB function 교체 여부를 확인해야 합니다.

### 롤백
- [ ] 플래그/설정으로 우회 가능
- [x] 배포 롤백(이전 태그/이미지) 가능
- [x] 데이터 롤백/역마이그레이션 필요 (절차: 추가 trigger/function 제거 또는 이전 function 재적용)

---

## 📸 증빙 (선택)
- 스크린샷/로그/메트릭 링크: 로컬 통합 테스트 PASS 로그
- 전/후 비교(가능하면): `to_char` 월 비교에서 month range query로 전환

---

## 💬 리뷰 가이드 (필수)
- 꼭 봐야 하는 파일/로직: `HomeSummaryService`, `HomeSummaryIntegrationTest`, `DomainSchemaSmokeTest`, `V3__home_trigger_test_hardening.sql`
- 트레이드오프/대안: 기존 V2 migration은 수정하지 않고 신규 V3 migration으로 분리했습니다.
- 성능/보안/호환성 영향: 홈 요약 월별 집계가 range 조건을 사용해 인덱스 활용 여지를 높입니다.

### CODE REVIEW

File: src/main/resources/db/migration/V4__home_trigger_test_hardening.sql [hunk 1]
```diff
@@ -0,0 +1,63 @@
+create or replace function prevent_self_check_response_set_delete()
```
Comment (ekdmssld) [inline src/main/resources/db/migration/V4__home_trigger_test_hardening.sql:1]: 현재 이 PR이 `V3__home_trigger_test_hardening.sql`을 추가하고 있는데, 열려 있는 PR #48도 `V3__notification_reminder_idempotency.sql`를 추가하고 있습니다. 중복인가요?

두 PR이 모두 develop에 머지되면 Flyway 기준으로 version 3 migration이 2개가 되어 `Found more than one migration with version 3` 형태로 애플리케이션 부팅/마이그레이션이 실패할 수 있어 보입니다.

둘 중 나중에 머지될 PR의 migration version을 `V4` 이상으로 조정하거나, 선행 머지 순서를 정해서 rebase 후 버전을 다시 잡는 게 필요합니다.

File: src/main/java/com/sagongsa/backend/home/HomeSummaryService.java [hunk 2]
```diff
@@ -177,6 +178,9 @@ private List<NotificationSummary> findLatestNotifications(UUID userId) {
 	}
 
 	private BigDecimal findRationalChoiceRate(UUID userId, String yearMonth, ZoneId zoneId) {
+		YearMonth targetMonth = YearMonth.parse(yearMonth);
+		OffsetDateTime monthStart = OffsetDateTime.from(targetMonth.atDay(1).atStartOfDay(zoneId));
```
Comment (ekdmssld) [inline src/main/java/com/sagongsa/backend/home/HomeSummaryService.java:182]: `to_char(decided_at at time zone ?, 'YYYY-MM')`를 range query로 바꾼 방향은 좋지만, 이 변경의 핵심은 사용자 timezone 기준 월 경계가 유지되는지라서, 테스트에 월 경계 케이스가 하나 있으면 더 안전할 것 같습니다.

예를 들어 UTC 기준으로는 전월/익월처럼 보이지만 `Asia/Seoul` 기준으로는 현재 월에 포함되는 decision, 또는 그 반대 케이스를 넣어서 `decided_at >= monthStart and decided_at < nextMonthStart`가 기존 계약과 동일하게 동작하는지 검증하면 좋겠습니다.

File: src/test/java/com/sagongsa/backend/home/HomeSummaryIntegrationTest.java
Comment (ekdmssld) [inline src/test/java/com/sagongsa/backend/home/HomeSummaryIntegrationTest.java:25]: 테스트 fixture가 현재 월 기준으로 정리된 건 좋은데 `BASE_TIME`은 class load 시점의 `YearMonth.now(SEOUL_ZONE)`이고, 각 테스트 안의 `yearMonth`는 실행 시점의 `YearMonth.now(SEOUL_ZONE)`라서 아주 드물게 월 경계를 걸치면 다시 어긋날 수 있습니다.

완전한 날짜 독립성을 목표로 한다면 테스트 안에서 기준 `YearMonth`/`BASE_TIME`을 하나의 변수로 만들거나, Clock 주입까지는 아니더라도 동일 기준값에서 파생시키는 식으로 맞추면 더 안정적일 것 같습니다.

**Review by ekdmssld — COMMENTED** [review]
홈 요약 rational choice 집계를 range query로 바꾼 방향과 trigger edge case를 smoke test로 고정한 점은 좋은거 같습니다. 수고하셨어요. 

migration version은 수정이 필요해 보입니다. 현재 이 PR이 `V3__home_trigger_test_hardening.sql`을 추가하고 있고, PR #48도 별도의 `V3__...sql` migration을 추가하고 있어서 둘 다 머지되면 Flyway duplicate version 문제가 날 수 있습니다. 머지 순서에 맞춰 migration version을 재조정해 주세요.

추가로 timezone 기준 월 경계 테스트와 HomeSummaryIntegrationTest의 기준 시간 단일화까지 보강되면 이번 PR의 “테스트 안정화” 목적이 더 확실해질 것 같습니다.

File: src/main/resources/db/migration/V4__home_trigger_test_hardening.sql (hunk 1 above)
Comment (PHJ2000) [reply src/main/resources/db/migration/V4__home_trigger_test_hardening.sql:1]: 반영했습니다. PR의 신규 DB migration 파일을 `V3`에서 `V4`로 변경해서 develop의 기존 V3와 충돌하지 않게 했습니다.

File: src/main/java/com/sagongsa/backend/home/HomeSummaryService.java (hunk 2 above)
Comment (PHJ2000) [reply src/main/java/com/sagongsa/backend/home/HomeSummaryService.java:182]: 반영했습니다. 한국 시간 월 경계에서 합리적 선택 비율이 현재 월 데이터만 집계되는 회귀 테스트를 추가했습니다. 월 시작 직전/월 시작/다음 달 시작 직전/다음 달 시작 케이스를 넣어서 조회 범위 경계를 확인하도록 했습니다.

File: src/test/java/com/sagongsa/backend/home/HomeSummaryIntegrationTest.java
Comment (PHJ2000) [reply src/test/java/com/sagongsa/backend/home/HomeSummaryIntegrationTest.java:25]: 반영했습니다. 테스트에서 현재 월 기준값을 한 번만 잡도록 정리해서, 예산 월과 결정 시각이 서로 다른 달로 갈라지지 않게 했습니다.
